### ISSUE TITLE
zeros_like / rand_like / randn_like fails on MPS for tensors loaded with pil / torchvision  (caused by `.permute`)

### ISSUE THREAD

**Issue by PiotrCzapla** [pytorch/pytorch#94190]
### 🐛 Describe the bug

If an image is loaded with pil/torch and put on `mps` device directly following by some operation (like multiplication) a  subsequent call to `*_like` fails. It outputs parts of the image or the image it self.

<img width="505" alt="Screenshot 2023-02-06 at 17 31 19" src="https://user-images.githubusercontent.com/340180/217028949-68cf95da-748a-403a-9397-97dbc64a7b47.png">

I've narrowed the issue to a call to `.permute` that transpose HWC to CHW, it seems that if such permuted tensor is put on a `mps` device and then it is used in a computation (`*`, `.sum`), a subsequent *_like call will fail. 
Here is a minimal example that shows the issue using `zeros_like`

```
import torch
shape=[3, 2, 2]

content_im = (torch.randn(shape).float())
content_im = content_im.permute((2, 0, 1))
content_im = content_im.to('mps')

content_im *1.0  # an operation on tensor is required to trigger the issue, it can be .sum() as well

print('zero_like on .contiguous()')
print(torch.zeros_like(content_im.contiguous()))

print('zero_like on permuted version') # this fails
print(torch.zeros_like(content_im))     # it output the content of content_im
```

Here is the output on torch 1.13.1:
```
zero_like on .contiguous()
/opt/homebrew/Caskroom/mambaforge/base/envs/miniai_stbl/lib/python3.10/site-packages/torch/_tensor_str.py:115: UserWarning: The operator 'aten::nonzero' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/_temp/anaconda/conda-bld/pytorch_1670525498485/work/aten/src/ATen/mps/MPSFallback.mm:11.)
  nonzero_finite_vals = torch.masked_select(
tensor([[[0., 0.],
         [0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.],
         [0., 0.]]], device='mps:0')
zero_like on permuted version
tensor([[[ 0.0000,  0.0000],
         [ 0.6181, -0.2015],
         [-0.3977, -0.0595]],

        [[ 0.0000, -0.5599],
         [ 0.8940, -1.0979],
         [-0.2532, -1.7495]]], device='mps:0')
```

The todays nightly ('2.0.0.dev20230206') build fails with assertion error:
```
zero_like on .contiguous()
tensor([[[0., 0.],
         [0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.],
         [0., 0.]]], device='mps:0')
zero_like on permuted version
loc("mps_slice_2"("(mpsFileLoc): /AppleInternal/Library/BuildRoots/9e200cfa-7d96-11ed-886f-a23c4f261b56/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm":349:0)): error: invalid start, length (2, 2) for size 3
LLVM ERROR: Failed to infer result type(s).
[1]    73352 abort      python nbs/mps_torchvision_issue.py
```

The  multiplied  view of an image is similar to the resnet issue  #86954 and fail caused by `permute` looks  similar to #88183, I've made a new issue to provide some evidence that this two might be linked.

### Versions

Collecting environment information...
PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.2 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.24.2
Libc version: N/A

Python version: 3.10.9 (main, Jan 11 2023, 09:18:18) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.2-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.5
[pip3] torch==1.13.1
[pip3] torchaudio==0.13.1
[pip3] torcheval==0.0.6
[pip3] torchtnt==0.0.5
[pip3] torchvision==0.14.1
[conda] numpy                     1.23.5          py310hb93e574_0  
[conda] numpy-base                1.23.5          py310haf87e8b_0  
[conda] pytorch                   1.13.1                 py3.10_0    pytorch
[conda] torchaudio                0.13.1                py310_cpu    pytorch
[conda] torcheval                 0.0.6                    pypi_0    pypi
[conda] torchtnt                  0.0.5                    pypi_0    pypi
[conda] torchvision               0.14.1                py310_cpu    pytorch

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by PiotrCzapla** [comment]
@razarmehr  I've compiled the version from PR and tested with the code above. The crash disappears however the result is still incorrect.  zeros_like returns a random tensor. Here is output from the test script.
```
zero_like on .contiguous()
tensor([[[0., 0.],
         [0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.],
         [0., 0.]]], device='mps:0')
zero_like on permuted version
tensor([[[ 0.0000e+00,  0.0000e+00],
         [-1.4609e+00,  1.4720e-01],
         [ 3.8663e-01, -3.2391e-01]],

        [[ 0.0000e+00,  1.5724e+00],
         [ 1.2158e+00, -5.9325e-04],
         [-7.7740e-02,  1.2826e+00]]], device='mps:0')
```

The second output should return tensors with zeros as well. I'm using version from the pr #94259
>>> torch.version.__version__
'2.0.0a0+gitf9b0131'

**Comment by malfet** [comment]
Reopening this one (though I wonder if it would make sense to update the title)
@PiotrCzapla just to check, output in 1.13 was also wrong, wasn't it?

**Comment by PiotrCzapla** [comment]
@malfet, you are correct the issue was discovered in 1.13. It was crashing with nightly build. After the PR The crash is fixed and we are back to the incorrect behaviour of 1.13. If you add the following line at the end of  `def test_permute_slicing(self):`
```
self.assertEqual(torch.zeros_like(cpu_out), torch.zeros_like(mps_out))
```
The test will fail. 

How would you like to update the title? I added a mention to pil / torchvision to make it easier to find it.  It took mea while to narrow this down to the `.permute` followed by computation.  Especially that the issue manifest it self for `randn_like` where it is easy to overlook.

**Comment by razarmehr** [comment]
Thanks @PiotrCzapla . I'll update the `test_permute_slicing()` with the `assertEqual()` code that you sent and will work on a fix.

**Comment by razarmehr** [comment]
This should be fixed with PR #94479.
